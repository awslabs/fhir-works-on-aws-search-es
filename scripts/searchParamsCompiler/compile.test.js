/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

const compile = require('./compile');

describe('compile', () => {
    test(`simple path - Patient.communication.language`, () => {
        const compiled = compile([
            {
                url: 'http://hl7.org/fhir/SearchParameter/Patient-language',
                name: 'language',
                type: 'token',
                description: 'Language code (irrespective of use value)',
                base: ['Patient'],
                expression: 'Patient.communication.language',
            },
        ]);

        expect(compiled).toMatchSnapshot();
    });

    test(`simple where - Library.relatedArtifact.where(type='predecessor').resource`, () => {
        const compiled = compile([
            {
                name: 'predecessor',
                description: 'What resource is being referenced',
                base: ['Library'],
                type: 'reference',
                expression: "Library.relatedArtifact.where(type='predecessor').resource",
                target: ['Account', 'ActivityDefinition'],
            },
        ]);
        expect(compiled).toMatchSnapshot();
    });

    test(`where with resolve() is - Person.link.target.where(resolve() is RelatedPerson)`, () => {
        const compiled = compile([
            {
                url: 'http://hl7.org/fhir/SearchParameter/Person-relatedperson',
                name: 'relatedperson',
                type: 'reference',
                description: 'The Person links to this RelatedPerson',
                base: ['Person'],
                expression: 'Person.link.target.where(resolve() is RelatedPerson)',
                target: ['RelatedPerson'],
            },
        ]);
        expect(compiled).toMatchSnapshot();
    });
    test(`as - (ConceptMap.source as uri)`, () => {
        const compiled = compile([
            {
                url: 'http://hl7.org/fhir/SearchParameter/ConceptMap-source-uri',
                name: 'source-uri',
                type: 'reference',
                description: 'The source value set that contains the concepts that are being mapped',
                base: ['ConceptMap'],
                expression: '(ConceptMap.source as uri)',
                target: ['ValueSet'],
            },
        ]);
        expect(compiled).toMatchSnapshot();
    });
    test(`OR operator - CapabilityStatement.title | CodeSystem.title | ConceptMap.title | ImplementationGuide.title | MessageDefinition.title | OperationDefinition.title | StructureDefinition.title | StructureMap.title | TerminologyCapabilities.title | ValueSet.title`, () => {
        const compiled = compile([
            {
                url: 'http://hl7.org/fhir/SearchParameter/conformance-title',
                name: 'title',
                type: 'string',
                description: 'Multiple Resources...',
                base: [
                    'CapabilityStatement',
                    'CodeSystem',
                    'ConceptMap',
                    'ImplementationGuide',
                    'MessageDefinition',
                    'OperationDefinition',
                    'StructureDefinition',
                    'StructureMap',
                    'TerminologyCapabilities',
                    'ValueSet',
                ],
                expression:
                    'CapabilityStatement.title | CodeSystem.title | ConceptMap.title | ImplementationGuide.title | MessageDefinition.title | OperationDefinition.title | StructureDefinition.title | StructureMap.title | TerminologyCapabilities.title | ValueSet.title',
            },
        ]);
        expect(compiled).toMatchSnapshot();
    });
    test(`OR operator with same base resource - (ExampleScenario.useContext.value as Quantity) | (ExampleScenario.useContext.value as Range)`, () => {
        const compiled = compile([
            {
                url: 'http://hl7.org/fhir/SearchParameter/ExampleScenario-context-quantity',
                name: 'context-quantity',
                type: 'quantity',
                description: 'A quantity- or range-valued use context assigned to the example scenario',
                base: ['ExampleScenario'],
                expression:
                    '(ExampleScenario.useContext.value as Quantity) | (ExampleScenario.useContext.value as Range)',
            },
        ]);
        expect(compiled).toMatchSnapshot();
    });
});
