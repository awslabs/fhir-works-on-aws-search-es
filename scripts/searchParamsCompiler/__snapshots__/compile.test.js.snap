// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile OR operator - CapabilityStatement.title | CodeSystem.title | ConceptMap.title | ImplementationGuide.title | MessageDefinition.title | OperationDefinition.title | StructureDefinition.title | StructureMap.title | TerminologyCapabilities.title | ValueSet.title 1`] = `
Object {
  "CapabilityStatement": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "CapabilityStatement",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "CodeSystem": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "CodeSystem",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "ConceptMap": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "ConceptMap",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "ImplementationGuide": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "ImplementationGuide",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "MessageDefinition": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "MessageDefinition",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "OperationDefinition": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "OperationDefinition",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "StructureDefinition": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "StructureDefinition",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "StructureMap": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "StructureMap",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "TerminologyCapabilities": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "TerminologyCapabilities",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
  "ValueSet": Object {
    "title": Object {
      "compiled": Array [
        Object {
          "path": "title",
          "resourceType": "ValueSet",
        },
      ],
      "description": "Multiple Resources...",
      "target": undefined,
      "type": "string",
    },
  },
}
`;

exports[`compile OR operator with same base resource - (ExampleScenario.useContext.value as Quantity) | (ExampleScenario.useContext.value as Range) 1`] = `
Object {
  "ExampleScenario": Object {
    "context-quantity": Object {
      "compiled": Array [
        Object {
          "path": "useContext.valueQuantity",
          "resourceType": "ExampleScenario",
        },
        Object {
          "path": "useContext.valueRange",
          "resourceType": "ExampleScenario",
        },
      ],
      "description": "A quantity- or range-valued use context assigned to the example scenario",
      "target": undefined,
      "type": "quantity",
    },
  },
}
`;

exports[`compile as - (ConceptMap.source as uri) 1`] = `
Object {
  "ConceptMap": Object {
    "source-uri": Object {
      "compiled": Array [
        Object {
          "path": "sourceUri",
          "resourceType": "ConceptMap",
        },
      ],
      "description": "The source value set that contains the concepts that are being mapped",
      "target": Array [
        "ValueSet",
      ],
      "type": "reference",
    },
  },
}
`;

exports[`compile simple path - Patient.communication.language 1`] = `
Object {
  "Patient": Object {
    "language": Object {
      "compiled": Array [
        Object {
          "path": "communication.language",
          "resourceType": "Patient",
        },
      ],
      "description": "Language code (irrespective of use value)",
      "target": undefined,
      "type": "token",
    },
  },
}
`;

exports[`compile simple where - Library.relatedArtifact.where(type='predecessor').resource 1`] = `
Object {
  "Library": Object {
    "predecessor": Object {
      "compiled": Array [
        Object {
          "condition": Array [
            "relatedArtifact.type",
            "=",
            "predecessor",
          ],
          "path": "relatedArtifact.resource",
          "resourceType": "Library",
        },
      ],
      "description": "What resource is being referenced",
      "target": Array [
        "Account",
        "ActivityDefinition",
      ],
      "type": "reference",
    },
  },
}
`;

exports[`compile where with resolve() is - Person.link.target.where(resolve() is RelatedPerson) 1`] = `
Object {
  "Person": Object {
    "relatedperson": Object {
      "compiled": Array [
        Object {
          "condition": Array [
            "link.target",
            "resolve",
            "RelatedPerson",
          ],
          "path": "link.target",
          "resourceType": "Person",
        },
      ],
      "description": "The Person links to this RelatedPerson",
      "target": Array [
        "RelatedPerson",
      ],
      "type": "reference",
    },
  },
}
`;
